### PR TITLE
Report a new location only after receiving a non-null update

### LINKED ISSUE #75: Stale location reported when backend calls report(null)
This came out of an issue in the openBmap backend, https://github.com/wish7code/org.openbmap.unifiedNlpProvider/issues/13: After getting stale locations under some circumstances, I did some investigations and introduced some modifications into the backend.

The backend does a combined cell/wifi lookup, either online or using an offline DB (all my tests were done with the offline DB).

Steps to reproduce:
1. Fire up an appropriate app in an environment with known wifis. I used SatStat, available on F-Droid, which visualizes both GPS and network location along with their accuracies. Check location, then return to home screen making sure no other app requests location updates).
1. Move to a location with no known cells and wifis at all, and fire up the app again.

The first step behaved as I expected: the modified backend gave me a crisp (+/- 20 m) location. The second step was performed some 5 minutes after leaving the previous location. SatStat claimed I was still within 20 meters of the old location, even after I killed and restarted it (to rule out any caching in SatStat).

Analyzing the logcat, I find that I originally had 26 wifis in range, from which the backend calculated a location and reported it. In the metro station, the backend reported that no wifis were in range and the only visible cell was not in the DB. It reported a null location, <del>resulting in a call to `report()` with a `null` argument</del>. Then, however, UnifiedNLP (specifically NlpLocationBackendFuser) reports the same location it has given out 5 minutes before, with the same et=+15d21h34m50s905ms (which I suppose is the timestamp) and the same accuracy acc=20.

Are backends supposed to call `report(null)` when they are unable to determine the current location, or should they simply not call `report()` at all? This should probably be documented in the API README.

In any case, the fuser probably should not give out old locations unmodified: when processing any location, its accuracy should be penalized based on its age (if you're interested, I can provide a link to my branch of the openBmap backend for an example on how to fuse locations of different ages) and the resulting location should have a recent timestamp.

### PR DESCRIPTION
**Pull request by mvglasow**
Prevents stale locations from being re-sent when backends supply null locations, fixes #75

### CODE REVIEW

**Comment by mar-v-in** [comment]
This does not allow to use forced locations (primarily used for debugging currently, but might also be useful for automated testing in the future).

Why don't you just verify that reportLocation is only called after the location was updated (i.e. location to be reported has newer timestamp than last report)?

**Comment by mar-v-in** [comment]
Additionally, report(null) is now different from report(old location) which should not be the case (no report, null report and old location report should all be treated the same as they all do mean that there is no new data available)

**Comment by mvglasow** [comment]
3b995b8 modifies `BackendHelper` to compare timestamps and report locations only if the new location has a more recent timestamp than the previous one.

For forced locations, eead0bd adds code to make them behave in a similar way as a location update from a backend: a forced location which is not more recent than the last forced location is silently discarded. When a new location is forced, `reportLocation()` gets called, resulting in a single location update. Differences to a backend: forcing a null location will clear the last forced location (which is the same behavior we've had up to now), and forced locations are always asynchronous (there is no equivalent to `update()` for forced locations).

File: unifiednlp-base/src/main/java/org/microg/nlp/location/BackendFuser.java [hunk 1]
```diff
@@ -78,12 +78,15 @@ public void bind() {
     }
 
     public void update() {
+        Boolean hasUpdates = false;
```
Comment (mar-v-in) [inline unifiednlp-base/src/main/java/org/microg/nlp/location/BackendFuser.java]: should be `boolean` instead of `Boolean`

**Comment by mvglasow** [comment]
Fixed in 1093f40

**Comment by n76** [comment]
Both of my backends are now causing a crash. From the trace back (below) it seems that I don't have the time set properly on the location report, but I am getting this even if I perform a myLocReport.setTime(System.currentTimeMillis()) immediately before calling report().

I've taken a look at your Apple and Mozilla backends to see what you are doing differently and it appears that the setTime(System.currentTimeMillis()) call is what you use.

Any ideas?

```
E/AndroidRuntime(22629): Process: org.fitchfamily.android.wifi_backend, PID: 22629
E/AndroidRuntime(22629): java.lang.RuntimeException: Error receiving broadcast Intent { act=android.net.wifi.SCAN_RESULTS flg=0x4000010 } in org.fitchfamily.android.wifi_backend.wifi.WifiReceiver@39f248db
E/AndroidRuntime(22629): 	at android.app.LoadedApk$ReceiverDispatcher$Args.run(LoadedApk.java:876)
E/AndroidRuntime(22629): 	at android.os.Handler.handleCallback(Handler.java:739)
E/AndroidRuntime(22629): 	at android.os.Handler.dispatchMessage(Handler.java:95)
E/AndroidRuntime(22629): 	at android.os.Looper.loop(Looper.java:135)
E/AndroidRuntime(22629): 	at android.app.ActivityThread.main(ActivityThread.java:5254)
E/AndroidRuntime(22629): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime(22629): 	at java.lang.reflect.Method.invoke(Method.java:372)
E/AndroidRuntime(22629): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:903)
E/AndroidRuntime(22629): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:698)
E/AndroidRuntime(22629): Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'long android.location.Location.getTime()' on a null object reference
E/AndroidRuntime(22629): 	at android.os.Parcel.readException(Parcel.java:1552)
E/AndroidRuntime(22629): 	at android.os.Parcel.readException(Parcel.java:1499)
E/AndroidRuntime(22629): 	at org.microg.nlp.api.LocationCallback$Stub$Proxy.report(LocationCallback.java:91)
E/AndroidRuntime(22629): 	at org.microg.nlp.api.LocationBackendService.report(LocationBackendService.java:52)
E/AndroidRuntime(22629): 	at org.fitchfamily.android.wifi_backend.backend.BackendService.process(BackendService.java:186)
E/AndroidRuntime(22629): 	at org.fitchfamily.android.wifi_backend.wifi.WifiReceiver.onReceive(WifiReceiver.java:92)
E/AndroidRuntime(22629): 	at android.app.LoadedApk$ReceiverDispatcher$Args.run(LoadedApk.java:866)
E/AndroidRuntime(22629): 	... 8 more
```

**Comment by mar-v-in** [comment]
@n76: Use v1.6.4

**Comment by n76** [comment]
I'm testing against GmsCore v0.2.2, listed as the latest in your downloads and just showing up in your F-Droid repository (I'm using F-Droid to keep up to date with your releases).

FWIW, I am seeing the same crashes and similar stack traces on your Apple and Mozilla backends.

**Comment by mar-v-in** [comment]
Please try to download the app again (or update the f-droid repo) there was a broken build online for a few minutes.

**Comment by n76** [comment]
Apparently that was the problem. . . Not crashing now. Thanks!

**Pull request merged**